pinot-core/src/main/java/org/apache/pinot/core/transport/AsyncQueryResponse.java

```diff
@@ -50,21 +52,31 @@ public class AsyncQueryResponse implements QueryResponse {
   private volatile ServerRoutingInstance _failedServer;
   private volatile Exception _exception;
 
-  public AsyncQueryResponse(QueryRouter queryRouter, long requestId, Set<ServerRoutingInstance> serversQueried,
-      long startTimeMs, long timeoutMs, ServerRoutingStatsManager serverRoutingStatsManager) {
+  public AsyncQueryResponse(QueryRouter queryRouter, long requestId,
+//      Set<ServerQueryRoutingContext> serverQueryRoutingContexts,
+      Map<ServerRoutingInstance, List<InstanceRequest>> requestMap, long startTimeMs, long timeoutMs,
+      ServerRoutingStatsManager serverRoutingStatsManager) {
     _queryRouter = queryRouter;
     _requestId = requestId;
-    int numServersQueried = serversQueried.size();
-    _responseMap = new ConcurrentHashMap<>(HashUtil.getHashMapCapacity(numServersQueried));
+    _responses = new ConcurrentHashMap<>();
     _serverRoutingStatsManager = serverRoutingStatsManager;
-    for (ServerRoutingInstance serverRoutingInstance : serversQueried) {
-      // Record stats related to query submission just before sending the request. Otherwise, if the response is
-      // received immediately, there's a possibility of updating query response stats before updating query
-      // submission stats.
-      _serverRoutingStatsManager.recordStatsForQuerySubmission(requestId, serverRoutingInstance.getInstanceId());
-      _responseMap.put(serverRoutingInstance, new ServerResponse(startTimeMs));
+    int numQueriesIssued = 0;
+    for (Map.Entry<ServerRoutingInstance, List<InstanceRequest>> serverRequests : requestMap.entrySet()) {
+      for (InstanceRequest request : serverRequests.getValue()) {
+        // Record stats related to query submission just before sending the request. Otherwise, if the response is
+        // received immediately, there's a possibility of updating query response stats before updating query
+        // submission stats.
+        _serverRoutingStatsManager.recordStatsForQuerySubmission(requestId, serverRequests.getKey().getInstanceId());
+
+        _responses.computeIfAbsent(serverRequests.getKey(), k -> new ConcurrentHashMap<>())
+            // we use query hash so that the same hash ID can be passed back from servers more easily than trying to
+            // instantiate a valid InstanceRequest obj and send its hash
+            .put(request.getQuery().getPinotQuery().hashCode(), new ServerResponse(startTimeMs));
```

Comment:
@jackjlli what are your thoughts on using the table name instead of the query hash as the key for the nested map here? The advantages would be that it's more "human understandable" as compared to an integer value. I believe that there should be no legitimate case where a single server needs to be sent multiple queries to the same _physical_ table name (i.e. with type), so table name should uniquely identify a query.  I also think it would make backward compatibility easier to support.

Thoughts? Any disadvantages you could foresee using table name as the key? How about multi-stage enging considerations?

Reply:
Right, it is possible that queries coming to multi-stage engine may query more than 1 table, and every server may not hold data for just 1 table though.

Reply:
Right, that's my understanding as well.  I _think_ that using physical table name (i.e. with type suffix) should be unique per-server-per-request-id.  My understanding is that for a given query issued to a broker, each server should only fetch 0 or 1 data table for a given physical table.  I'm under the impression that there is no legitimate case where the same server for the same request_id would receive more than 1 query for the same physical table.

Is that accurate?  If so, I believe that using the physical table name as the hash key will be highly preferable over query hash.

Reply:
I see your point, yes it should be okay to just use the physical table name here, as each thread will just work on 1 request. 👍